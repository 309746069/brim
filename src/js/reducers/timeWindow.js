import createReducer from "./createReducer"
import {createSelector} from "reselect"
import {getCurrentSpaceTimeWindow} from "./spaces"
import {isTimeWindow} from "../models/TimeWindow"
import isArray from "lodash/isArray"
import isDate from "lodash/isDate"

const initialState = {
  inner: null,
  outer: null
}

export default createReducer(initialState, {
  OUTER_TIME_WINDOW_SET: (state, {timeWindow}) => ({
    ...state,
    outer: timeWindow
  }),
  INNER_TIME_WINDOW_SET: (state, {timeWindow}) => ({
    ...state,
    inner: timeWindow
  })
})

export const getRawOuterTimeWindow = state => state.timeWindow.outer
export const getRawInnerTimeWindow = state => state.timeWindow.inner
export const getOuterTimeWindow = createSelector(getRawOuterTimeWindow, value =>
  makeDates(value)
)
export const getInnerTimeWindow = createSelector(getRawInnerTimeWindow, value =>
  makeDates(value)
)
export const getTimeWindow = createSelector(
  getOuterTimeWindow,
  getCurrentSpaceTimeWindow,
  (userTimeWindow, spaceTimeWindow) =>
    isTimeWindow(userTimeWindow) ? userTimeWindow : spaceTimeWindow
)

const makeDates = value =>
  isArray(value)
    ? value.map(item => (isDate(item) ? item : new Date(item)))
    : value
