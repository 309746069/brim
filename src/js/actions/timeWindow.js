/* @flow */

import {getCurrentSpaceTimeWindow} from "../reducers/spaces"
import type {TimeWindow} from "../reducers/timeWindow"
import type {DateTuple} from "../lib/TimeWindow"

export const setOuterTimeWindow = (timeWindow: DateTuple) => ({
  type: "OUTER_TIME_WINDOW_SET",
  timeWindow
})

export const setInnerTimeWindow = (timeWindow: ?DateTuple) => ({
  type: "INNER_TIME_WINDOW_SET",
  timeWindow
})

export const setOuterFromTime = (date: Date) => ({
  type: "OUTER_FROM_TIME_SET",
  date
})

export const setOuterToTime = (date: Date) => ({
  type: "OUTER_TO_TIME_SET",
  date
})

export const clearTimeWindows = () => ({
  type: "TIME_WINDOWS_CLEAR"
})

export const restoreTimeWindow = (value: TimeWindow) => ({
  type: "TIME_WINDOW_RESTORE",
  value
})

export const init = () => (dispatch: Function, getState: Function) => {
  const timeWindow = getCurrentSpaceTimeWindow(getState())
  dispatch(setOuterTimeWindow(timeWindow))
}
