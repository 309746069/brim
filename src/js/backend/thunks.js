/* @flow */
import type {Cluster} from "../state/clusters/types"
import type {Span} from "../BoomClient/types"
import type {Thunk} from "../state/types"
import {ZqVersionError} from "../models/Errors"
import {createError} from "../state/errors"
import {getCurrentSpaceName} from "../state/reducers/spaces"
import ErrorFactory from "../models/ErrorFactory"
import brim from "../brim"
import notice from "../state/notice"
import search from "../state/search"

export function fetchSearch(program: string, span: Span, space: string): Thunk {
  return (dispatch, g, boom) => {
    dispatch(notice.clear())
    return boom
      .search(program, {searchSpan: span, searchSpace: space})
      .error((e) => handleError(e, dispatch))
  }
}

export function fetchSpaces() {
  return promise((boom) => boom.spaces.list())
}

export function fetchSpace(name: string) {
  return promise((boom) => boom.spaces.get(name))
}

export function createSpace(name: string): Thunk {
  return promise((boom) => boom.spaces.create({name}))
}

export function testConnection(cluster: Cluster): Thunk {
  return function(dispatch, _, boom) {
    boom.setOptions({
      host: cluster.host,
      port: parseInt(cluster.port),
      username: cluster.username,
      password: cluster.password
    })
    return dispatch(fetchSpaces())
  }
}

export function checkVersions(): Thunk {
  return function(dispatch, _, boom) {
    let cv = boom.clientVersion()
    boom.serverVersion().done((sv) => {
      if (cv.zq !== sv.zq) {
        dispatch(notice.set(new ZqVersionError({client: cv.zq, server: sv.zq})))
      }
    })
  }
}

export function inspectSearch(zql: string): Thunk {
  return function(_, getState, boom) {
    let [from, to] = search.getSpan(getState())
    let searchSpan = [brim.time(from).toDate(), brim.time(to).toDate()]
    let searchSpace = getCurrentSpaceName(getState())
    try {
      return boom.inspectSearch(zql, {searchSpan, searchSpace})
    } catch {
      return null
    }
  }
}

function promise(request): Thunk {
  return function(dispatch, _, boom) {
    return new Promise((resolve, reject) => {
      request(boom)
        .done((...args) => {
          dispatch(notice.clear())
          resolve(...args)
        })
        .error((e) => {
          handleError(e, dispatch)
          reject(e)
        })
    })
  }
}

function handleError(e, dispatch) {
  dispatch(notice.set(ErrorFactory.create(e)))
  dispatch(createError(e))
}
