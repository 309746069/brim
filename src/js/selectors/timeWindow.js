import {createSelector} from "reselect"
import {getCurrentSpace} from "../reducers/spaces"
import {toDate} from "../cast"
import {isTimeWindow} from "../models/TimeWindow"

export function getUserSelectedTimeWindow(state) {
  return state.timeWindow
}

export const getCurrentSpaceTimeWindow = createSelector(
  getCurrentSpace,
  space => {
    if (!space) return []
    const to = toDate(space.max_time.sec + "." + space.max_time.ns)
    const from = toDate(space.min_time.sec + "." + space.min_time.ns)
    return [from, to]
  }
)

export const getTimeWindow = createSelector(
  getUserSelectedTimeWindow,
  getCurrentSpaceTimeWindow,
  (userTimeWindow, spaceTimeWindow) =>
    isTimeWindow(userTimeWindow) ? userTimeWindow : spaceTimeWindow
)
