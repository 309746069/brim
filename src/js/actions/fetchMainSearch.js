import * as actions from "."
import * as selectors from "../selectors"
import * as outMessages from "../boom/outMessages"
import eventsReceiver from "../receivers/eventsReceiver"
import countByTimeReceiver from "../receivers/countByTimeReceiver"
import analyticsReceiver from "../receivers/analyticsReceiver"
import statsReceiver from "../receivers/statsReceiver"

export function fetchMainSearch({saveToHistory = true} = {}) {
  return (dispatch, getState, api) => {
    const query = selectors.getMainSearchQuery(getState())

    if (!query.isValid()) {
      console.warn("invalide query", query.toString())
      return
    }

    if (saveToHistory) {
      dispatch(
        actions.pushSearchHistory(selectors.getSearchHistoryEntry(getState()))
      )
    }

    dispatch(actions.requestMainSearch({saveToHistory}))

    if (query.hasAnalytics()) {
      api
        .send(outMessages.fetchMainSearch(query))
        .each(statsReceiver(dispatch))
        .channel(0, analyticsReceiver(dispatch, 0))
        .done(() => dispatch(actions.completeMainSearch()))
    } else {
      api
        .send(outMessages.fetchMainSearch(query))
        .each(statsReceiver(dispatch))
        .channel(1, eventsReceiver(dispatch))
        .channel(0, countByTimeReceiver(dispatch))
        .done(() => dispatch(actions.completeMainSearch()))
    }
  }
}
// Get the type of search we are performing
// If it has analytics
//   Add the analytics handler
// If it has events
//   Add the events handler
//   Add the analytics
//   Add the count by time
