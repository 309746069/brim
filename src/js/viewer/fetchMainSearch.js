/* @flow */

import type {Thunk} from "../state/types"
import {cancelSearchesByTag} from "../searches/cancelSearch"
import {clearViewer} from "../state/viewer/actions"
import {
  getInnerTimeWindow,
  getOuterTimeWindow
} from "../state/reducers/timeWindow"
import {getSearchProgram} from "../state/selectors/searchBar"
import {getSearchRecord} from "../state/selectors/searchRecord"
import {issueSearch} from "../searches/issueSearch"
import {recordSearch} from "../state/actions"
import {updateTab} from "../state/thunks/view"
import {validateProgram} from "../state/thunks/searchBar"
import SearchTemplateFactory from "../searches/SearchTemplateFactory"

type Options = {
  saveToHistory: boolean
}

export const fetchMainSearch = ({
  saveToHistory = true
}: Options = {}): Thunk => (dispatch, getState) => {
  const state = getState()
  if (!dispatch(validateProgram())) return
  dispatch(updateTab(state))
  if (saveToHistory) dispatch(recordSearch(getSearchRecord(state)))
  dispatch(cancelSearchesByTag("viewer"))
  dispatch(clearViewer())
  new SearchTemplateFactory({
    program: getSearchProgram(state),
    innerSpan: getInnerTimeWindow(state),
    outerSpan: getOuterTimeWindow(state),
    saveToHistory
  })
    .getTemplates()
    .forEach((template) => dispatch(issueSearch(template)))
}
