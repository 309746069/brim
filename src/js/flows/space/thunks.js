/* @flow */

import {NoSpacesError} from "../../models/Errors"
import type {Thunk} from "../../state/types"
import {clearViewer} from "../../state/viewer/actions"
import {fetchSpace, fetchSpaces} from "../../services/boom"
import {getCurrentSpaceTimeWindow} from "../../state/reducers/spaces"
import {killAllSearches} from "../../searches/cancelSearch"
import {
  setCurrentSpaceName,
  setSpaceInfo,
  setSpaceNames
} from "../../state/actions"
import brim from "../../brim"
import modal from "../../state/modal"
import notice from "../../state/notice"
import search from "../../state/search"
import submitSearch from "../submitSearch"

export function initSpace(space: string): Thunk {
  return function(dispatch, getState) {
    dispatch(killAllSearches())
    return dispatch(fetchSpaces()).then((spaces) => {
      if (spaces.length === 0) {
        dispatch(notice.set(new NoSpacesError()))
      } else {
        let name = spaces.includes(space) ? space : spaces[0]
        dispatch(fetchSpace(name)).then((info) => {
          let space = brim.space(info)

          dispatch(setCurrentSpaceName(name))
          dispatch(setSpaceInfo(info))

          if (space.empty()) {
            dispatch(clearViewer())
            dispatch(modal.show("nodata"))
          } else {
            const [_, max] = getCurrentSpaceTimeWindow(getState())
            dispatch(
              search.setSpanArgs([
                brim
                  .time(max)
                  .subtract(30, "minutes")
                  .toTs(),
                max
              ])
            )
            dispatch(submitSearch())
          }
        })
      }
    })
  }
}

export const refreshSpaces = (): Thunk => (dispatch) => {
  return dispatch(fetchSpaces()).then((spaces) =>
    dispatch(setSpaceNames(spaces))
  )
}
