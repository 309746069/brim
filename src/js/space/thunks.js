/* @flow */
import {NoSpacesError} from "../models/Errors"
import type {Thunk} from "../state/types"
import {clearViewer} from "../state/viewer/actions"
import {fetchSpace, fetchSpaces} from "../backend/thunks"
import {getCurrentSpaceTimeWindow} from "../state/reducers/spaces"
import {killAllSearches} from "../searches/cancelSearch"
import {setBackendError} from "../backend"
import {
  setCurrentSpaceName,
  setSpaceInfo,
  setSpaceNames
} from "../state/actions"
import {setOuterTimeWindow} from "../state/span/actions"
import {submitSearchBar} from "../state/thunks/searchBar"
import brim from "../brim"
import modal from "../modal"

export function initSpace(space: string): Thunk {
  return function(dispatch, getState) {
    dispatch(killAllSearches())
    return dispatch(fetchSpaces()).then((spaces) => {
      if (spaces.length === 0) {
        dispatch(setBackendError(new NoSpacesError()))
      } else {
        let name = spaces.includes(space) ? space : spaces[0]
        dispatch(fetchSpace(name)).then((info) => {
          let space = brim.space(info)

          dispatch(setCurrentSpaceName(name))
          dispatch(setSpaceInfo(info))

          if (space.empty()) {
            dispatch(clearViewer())
            dispatch(modal.show("nodata"))
          } else {
            const [_, max] = getCurrentSpaceTimeWindow(getState())
            dispatch(
              setOuterTimeWindow([
                brim
                  .time(max)
                  .subtract(30, "minutes")
                  .toDate(),
                max
              ])
            )
            dispatch(submitSearchBar())
          }
        })
      }
    })
  }
}

export const refreshSpaces = (): Thunk => (dispatch) => {
  return dispatch(fetchSpaces()).then((spaces) =>
    dispatch(setSpaceNames(spaces))
  )
}
