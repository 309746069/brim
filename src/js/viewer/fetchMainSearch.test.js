/* @flow */

import {
  changeSearchBarInput,
  setCurrentSpaceName,
  setInnerTimeWindow,
  setSpaceInfo
} from "../state/actions"
import {fetchMainSearch} from "./fetchMainSearch"
import {initTimeWindow} from "../state/thunks/timeWindow"
import MockBoomClient from "../test/MockBoomClient"
import initTestStore from "../test/initTestStore"

let store, boom

beforeEach(() => {
  boom = new MockBoomClient({host: "localhost", port: 123})
  store = initTestStore(boom)
})

const spaceInfo = {
  name: "ranch-logs",
  min_time: {sec: 1425564900, ns: 0},
  max_time: {sec: 1428917793, ns: 750000000}
}

test("fetching a regular search", () => {
  boom.stub("search")
  store.dispatchAll([
    setSpaceInfo(spaceInfo),
    setCurrentSpaceName("ranch-logs"),
    initTimeWindow(),
    changeSearchBarInput("_path=conn")
  ])

  store.clearActions()
  store.dispatch(fetchMainSearch())

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("not saving a search to history", () => {
  boom.stub("search")
  store.dispatchAll([
    setSpaceInfo(spaceInfo),
    setCurrentSpaceName("ranch-logs"),
    changeSearchBarInput("_path=conn")
  ])

  store.clearActions()
  store.dispatch(fetchMainSearch({saveToHistory: false}))

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an analytic search", () => {
  boom.stub("search")
  store.dispatchAll([
    setSpaceInfo(spaceInfo),
    setCurrentSpaceName("ranch-logs"),
    initTimeWindow(),
    changeSearchBarInput("_path=conn | count()")
  ])

  store.clearActions()
  store.dispatch(fetchMainSearch())

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an analytic search without history", () => {
  boom.stub("search")
  store.dispatchAll([
    setSpaceInfo(spaceInfo),
    setCurrentSpaceName("ranch-logs"),
    initTimeWindow(),
    changeSearchBarInput("_path=conn | count()")
  ])

  store.clearActions()
  store.dispatch(fetchMainSearch({saveToHistory: false}))

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an zoom search", () => {
  boom.stub("search")
  store.dispatchAll([
    setSpaceInfo(spaceInfo),
    setCurrentSpaceName("ranch-logs"),
    initTimeWindow(),
    setInnerTimeWindow([new Date(0), new Date(1)]),
    changeSearchBarInput("_path=conn | count()")
  ])

  store.clearActions()
  store.dispatch(fetchMainSearch())

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("fetching an zoom search without history", () => {
  boom.stub("search")
  store.dispatchAll([
    setSpaceInfo(spaceInfo),
    setCurrentSpaceName("ranch-logs"),
    initTimeWindow(),
    setInnerTimeWindow([new Date(0), new Date(1)]),
    changeSearchBarInput("_path=conn | count()")
  ])

  store.clearActions()
  store.dispatch(fetchMainSearch({saveToHistory: false}))

  expect(store.getActions().map((a) => a.type)).toMatchSnapshot()
})

test("a bad search query", () => {
  const actions = [
    setSpaceInfo(spaceInfo),
    setCurrentSpaceName("ranch-logs"),
    changeSearchBarInput("_ath="),
    fetchMainSearch()
  ]
  actions.forEach(store.dispatch)

  expect(store.getActions().map((action) => action.type)).toEqual(
    expect.arrayContaining(["SEARCH_BAR_PARSE_ERROR"])
  )
})
