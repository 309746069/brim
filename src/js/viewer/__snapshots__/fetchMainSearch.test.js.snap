// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetching a regular search 1`] = `
Array [
  "SHOW_LOGS_TAB",
  "SEARCH_HISTORY_PUSH",
  "VIEWER_CLEAR",
  "SEARCH_REGISTER",
  "SEARCH_REGISTER",
]
`;

exports[`fetching an analytic search 1`] = `
Array [
  "SHOW_ANALYTICS_TAB",
  "SEARCH_HISTORY_PUSH",
  "VIEWER_CLEAR",
  "SEARCH_REGISTER",
]
`;

exports[`fetching an analytic search without history 1`] = `
Array [
  "SHOW_ANALYTICS_TAB",
  "VIEWER_CLEAR",
  "SEARCH_REGISTER",
]
`;

exports[`fetching an zoom search 1`] = `
Array [
  "SHOW_ANALYTICS_TAB",
  "SEARCH_HISTORY_PUSH",
  "VIEWER_CLEAR",
  "SEARCH_REGISTER",
]
`;

exports[`fetching an zoom search without history 1`] = `
Array [
  "SHOW_ANALYTICS_TAB",
  "VIEWER_CLEAR",
  "SEARCH_REGISTER",
]
`;

exports[`not saving a search to history 1`] = `
Array [
  "SHOW_LOGS_TAB",
  "VIEWER_CLEAR",
  "SEARCH_REGISTER",
  "SEARCH_REGISTER",
]
`;
